base/inc/AdePT/TrackManager.h

```diff
@@ -0,0 +1,263 @@
+    const size_t TracksSize = sizeof(Track) * fCapacity;
+    COPCORE_CUDA_CHECK(cudaMalloc(&fInstance_d, sizeof(TrackManager<Track>)));
+    COPCORE_CUDA_CHECK(cudaMalloc(&fActiveTracks, QueueSize));
+    COPCORE_CUDA_CHECK(cudaMalloc(&fNextTracks, QueueSize));
+    COPCORE_CUDA_CHECK(cudaMalloc(&fBuffer, TracksSize));
+
+    device_impl_trackmgr::construct_trackmanager<Track>
+        <<<1, 1>>>(fInstance_d, fCapacity, fActiveTracks, fNextTracks, fBuffer);
+    return fInstance_d;
+#endif
+  }
+
+  void FreeFromDevice()
+  {
+#ifdef COPCORE_CUDA_COMPILER
+    COPCORE_CUDA_CHECK(cudaFree(fBuffer));
+    COPCORE_CUDA_CHECK(cudaFree(fActiveTracks));
+    COPCORE_CUDA_CHECK(cudaFree(fNextTracks));
+    COPCORE_CUDA_CHECK(cudaFree(fInstance_d));
+#endif
+  }
+
+  /// @brief Swap active and next track slots. Compact if the fill percentage is higher than the threshold.
+  /// @details Must be called after the stats were updated on host.
+  template <typename Stream>
+  bool SwapAndCompact(float compact_threshold, Stream stream)
+  {
+#ifndef COPCORE_CUDA_COMPILER
+    COPCORE_EXCEPTION("SwapAndCompact cannot be called in hos-only code");
+    return false;
+#else
+    // check if the compacting threshold is hit
+    int used     = fStats.GetNused();
+    int inFlight = fStats.fInFlight;
+    assert(used >= 0 && used < fCapacity);
+    // Estimate maximum space needed if we DON'T compress now
+    int needed = used + 2 * inFlight;
+    if (needed < compact_threshold * fCapacity || needed > fCapacity) {
+      device_impl_trackmgr::swap_active<Track><<<1, 1, 0, stream>>>(fInstance_d);
+      if (needed > fCapacity) {
+        std::cout << "TrackManager::SwapAndCompact  ALERT: not enough space left to compress " << used
+                  << " used slots. Consider increasing TrackManager capacity.\n";
+      }
+      return false;
+    }
+
+    constexpr int maxBlocks = 1024;
+    constexpr int threads   = 32;
+
+    int blocks = (inFlight + threads - 1) / threads;
+    blocks     = min(blocks, maxBlocks);
+
+    fStats.fStart = fStats.fNextStart & (fCapacity - 1);
+    int next_free = fStats.fStart + inFlight;
+
+    // printf("compacting %d / %d -> %d at slot: %d, next_free: %d\n", used, fCapacity, inFlight, fStats.fStart,
+    // next_free & (fCapacity - 1));
+    device_impl_trackmgr::defragment_buffer<Track>
+        <<<blocks, threads, 0, stream>>>(fInstance_d, inFlight, fStats.fStart);
+    device_impl_trackmgr::adjust_indices<Track><<<1, 1, 0, stream>>>(fInstance_d, fStats.fStart, next_free);
+    COPCORE_CUDA_CHECK(cudaStreamSynchronize(stream));
+    return true;
+#endif
+  }
+
+  /// @brief Host static call to clear the container.
+  template <typename Stream>
+  void Clear(Stream stream)
+  {
+    fStats.fStart     = 0;
+    fStats.fNextStart = 0;
+    fStats.fInFlight  = 0;
+#ifdef COPCORE_CUDA_COMPILER
+    device_impl_trackmgr::clear_trackmanager<Track><<<1, 1, 0, stream>>>(fInstance_d);
+#endif
```

Comment:
Hrm, this is is *really* bad because the compiler will generate different code depending on where this is called and then the linker gets to decide which version to choose. We already had this problem in the past, please don't get us back to that...

Reply:
Yes, I initially thought using the utility host and device, but dropped this. I'll make the header `.cuh` removing the unnecessary ifdefs